cache/lru/lru.go

```diff
@@ -0,0 +1,173 @@
+package lru
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/neutrino/cache"
+)
+
+var (
+	// ErrElementNotFound is returned when element isn't found in the cache.
+	ErrElementNotFound = fmt.Errorf("unable to find element")
+)
+
+// elementMap is an alias for a map from a generic interface to a list.Element.
+type elementMap map[interface{}]*list.Element
+
+// entry represents a (key,value) pair entry in the Cache. The Cache's list
+// stores entries which let us get the cache key when an entry is evicted.
+type entry struct {
+	key   interface{}
+	value cache.Value
+}
+
+// Cache provides a generic thread-safe lru cache that can be used for
+// storing filters, blocks, etc.
+type Cache struct {
+	// capacity represents how much this cache can hold. It could be number of
+	// elements or a number of bytes, decided by the cache.Value's Size.
+	capacity uint64
+
+	// size represents the size of all the elements currenty in the cache.
+	size uint64
+
+	// ll is a doubly linked list which keeps track of recency of used elements
+	// by moving them to the front.
+	ll *list.List
+
+	// cache is a generic cache which allows us to find an elements position
+	// in the ll list from a given key.
+	cache elementMap
+
+	// mtx is used to make sure the Cache is thread-safe.
+	mtx sync.RWMutex
+}
+
+// NewCache return a cache with specified capacity, the cache's size can't
+// exceed that given capacity.
+func NewCache(capacity uint64) *Cache {
+	return &Cache{
+		capacity: capacity,
+		ll:       list.New(),
+		cache:    make(elementMap),
+	}
+}
+
+// evict will evict as many elements as necessary to make enough space for a new
+// filter to be inserted. sizeTotal is the amount of space evict needs to make.
+// It could end up evicting more than sizeTotal if the size of the least
+// recently used element to evict was bigger than the size needed to evict.
+func (c *Cache) evict(sizeTotal uint64) error {
```

Comment:
Letting `evict` take the size of the element to insert (see comment below), it would go something like this:
```go
func (c *Cache) evict(needed uint64) error {
    room := c.capacity - c.size
    for room < needed {
        ... remove one element ...
    }
```